server/modules/category/service/categoryImportJob.js

```diff
@@ -0,0 +1,205 @@
+const Job = require('../../../job/job')
+
+const Survey = require('../../../../common/survey/survey')
+const Category = require('../../../../common/survey/category')
+const CategoryImportSummary = require('../../../../common/survey/categoryImportSummary')
+const CategoryLevel = require('../../../../common/survey/categoryLevel')
+const CategoryItem = require('../../../../common/survey/categoryItem')
+const Validator = require('../../../../common/validation/validator')
+const ValidatorErrorKeys = require('../../../../common/validation/validatorErrorKeys')
+const StringUtils = require('../../../../common/stringUtils')
+
+const SurveyManager = require('../../survey/manager/surveyManager')
+const CategoryManager = require('../manager/categoryManager')
+const CategoryImportCSVParser = require('./categoryImportCSVParser')
+const CategoryImportJobParams = require('./categoryImportJobParams')
+
+class CategoryImportJob extends Job {
+
+  constructor (params) {
+    super(CategoryImportJob.type, params)
+
+    this.itemsByCodes = {} // cache of category items by ancestor codes
+  }
+
+  async execute () {
+    // levels
+    let category = await this._importLevels()
+
+    // items
+    await this._importItems(category)
+
+    if (this.hasErrors()) {
+      // errors found
+      this.logDebug(`${Object.keys(this.errors).length} errors found`)
+      await this.setStatusFailed()
+    } else {
+      // no errors found
+      const surveyId = this.getSurveyId()
+      const user = this.getUser()
+
+      const items = Object.values(this.itemsByCodes)
+      await CategoryManager.insertItems(user, surveyId, items, this.tx)
+
+      this.logDebug(`${items.length} items imported`)
+    }
+  }
+
+  async _importLevels () {
+    this.logDebug('importing levels')
+
+    const categoryUuid = CategoryImportJobParams.getCategoryUuid(this.params)
+    const summary = CategoryImportJobParams.getSummary(this.params)
+
+    const surveyId = this.getSurveyId()
+    const user = this.getUser()
+
+    await CategoryManager.deleteLevelsByCategory(user, surveyId, categoryUuid, this.tx)
+    let category = await CategoryManager.fetchCategoryByUuid(surveyId, categoryUuid, true, false, this.tx)
```

Comment:
can the ```CategoryManager.deleteLevelsByCategory``` return the category? you won't need the fetch in that case

Reply:
Right

Reply:
ok right, but it still fetches after delete

Reply:
I'll do it... now ;)